### PR TITLE
Store: Move products fetching / listing to the data layer

### PR DESCRIPTION
**Pull request by justinshreve**
This PR moves the products fetching code from thunks to the data layer, similar to what was done in #16317. With the wpcom-http data layer, we'll get automatic retries for GET requests for free. 

It also adds some error handling to the product list page if we are unable to load the products correctly after those retries. Previously, nothing would happen -- now the user will get a message on the page instead of the list table.

I am committing these handlers to the old file location. #16444 moves them.

To Test:
* Run `npm run test-client client/extensions/woocommerce/state` and make sure all tests pass.
* Go to `http://calypso.localhost:3000/store/products/:site` and make sure that your products load correctly. Page between products as well.
* To test the error message, you can temporarily de-activate the `wc-api-dev` plugin.

### CODE REVIEW

File: client/extensions/woocommerce/state/data-layer/products/index.js [hunk 1]
```diff
@@ -75,3 +83,22 @@ export function handleProductRequest( { dispatch }, action ) {
 	const updatedSuccessAction = updatedAction( siteId, action, successAction );
 	dispatch( get( siteId, 'products/' + productId, updatedSuccessAction, failureAction ) );
 }
+
+export function handleProductsRequestSuccess( store, action, next, response ) {
+	const { body, status } = response.data;
+
+	// We needed to envelope the reponse to get the number of products and pages. see handleProductsRequest
+	if ( status !== 200 ) {
+		action.meta.dataLayer.error = { code: ( body.code || status ), message: body.message };
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Unfortunately, it looks like reaching into `action.meta` from the reducer is not recommended. That stance is taken here in another PR: https://github.com/Automattic/wp-calypso/pull/16385#issuecomment-317588993

I'll also be updating my existing product category PR to no longer use this functionality.

Comment (justinshreve) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: I'll set this back to `In Progress` for now, pending our chat. If I remove the last commit, this uses the `*_RECEIVE` pattern, but I'll hold off doing anything until we figure out what direction we want to go in.

File: client/extensions/woocommerce/app/products/index.js [hunk 2]
```diff
@@ -81,9 +84,16 @@ class Products extends Component {
...
 
 	render() {
-		const { className, site, translate, productsLoading, productsLoaded, totalProducts } = this.props;
+		const { className, site, translate, productsLoading, productsLoaded, totalProducts, productsError } = this.props;
 		const classes = classNames( 'products__list', className );
 
+		if ( productsError ) {
+			return <EmptyContent
+				title={ translate( 'We are having trouble loading your products.' ) }
+				line={ translate( 'Please try again.' ) }
```
Comment (coderkevin) [inline client/extensions/woocommerce/app/products/index.js]: I wonder what the user will interpret as "trying again". Are they expected to refresh the page in the browser? Or somehow refresh the API call?

File: client/extensions/woocommerce/state/data-layer/products/index.js [hunk 3]
```diff
@@ -75,3 +84,46 @@ export function handleProductRequest( { dispatch }, action ) {
 	const updatedSuccessAction = updatedAction( siteId, action, successAction );
 	dispatch( get( siteId, 'products/' + productId, updatedSuccessAction, failureAction ) );
 }
+
+export function handleProductsRequestSuccess( store, action, next, response ) {
+	const { siteId, page } = action;
+	const { headers, body, status } = response.data;
+
+	// We needed to envelope the reponse to get the number of products and pages. see handleProductsRequest
+	if ( status !== 200 ) {
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Why would this ever get back something other than a 200 and fire `handleProductsRequestSuccess`? If it's not successful, shouldn't it fire `handleProductsRequestError` instead? Or is another 2XX code possible that you're checking for?

File: client/extensions/woocommerce/state/sites/products/reducer.js [hunk 4]
```diff
@@ -81,15 +88,12 @@ export function productsDeleteSuccess( state, action ) {
...
 
-export function productsRequest( state, action ) {
-	const prevState = state || {};
-	const isLoading = setLoading( prevState, action.page, true );
-	return { ...prevState, isLoading };
-}
-
-export function productsRequestFailure( state, action ) {
+export function productsRequest( state, { page, meta: { dataLayer: { error, data } } } ) {
+	if ( data || error ) {
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/products/reducer.js]: I'm not certain this is what we should be looking for. In fact, from my understanding of the data-layer, this action should only get passed to the reducers once, the first time. After that it shouldn't be dispatched down further than the data-layer handlers. It kind of defeats the purpose of simplifying actions in the data layer if we have to guard for meta values here.

What that may mean is that we need to remove `next( action );` from the handler code to ensure that the action doesn't make it down to the reducer and only the `RECEIVE` action does.

File: client/extensions/woocommerce/state/sites/products/reducer.js [hunk 5]
```diff
@@ -139,3 +143,9 @@ function setLoading( state, page, newStatus ) {
 	isLoading[ page ] = newStatus;
 	return isLoading;
 }
+
+function setError( state, page, newStatus ) {
+	const isError = state.isError && { ...state.isError } || {};
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/products/reducer.js]: I think `{ ...undefined }` should return {}, right? So I think you can remove the && and || here.

File: client/extensions/woocommerce/state/sites/products/reducer.js [hunk 6]
```diff
@@ -139,3 +143,9 @@ function setLoading( state, page, newStatus ) {
 	isLoading[ page ] = newStatus;
 	return isLoading;
 }
+
+function setError( state, page, newStatus ) {
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/products/reducer.js]: I'll be honest, I'm not a huge fan of holding error state for each page load individually like this, but I can't think of a simpler way to do this better right now. I think it's okay to do it this way for now, but perhaps we should be thinking of a way to improve/consolidate code like this in the future.

File: client/extensions/woocommerce/state/data-layer/products/index.js (hunk 3 above)
Comment (justinshreve) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: I can make the comment above it a bit clearer, but we have to use https://developer.wordpress.org/rest-api/using-the-rest-api/global-parameters/#_envelope to get back the number of products and pagination pages from the API. The headers that get returned to the data layer are from `public-api.wordpress.com`/ the proxy.

**Comment by justinshreve** [comment]
Setting this `In Progress` until I return from AFK and can properly address everything. It's not attached to a current bug or issue, so it can wait to be merged.

**Pull request closed without merging**